src/main/java/net/ccbluex/liquidbounce/injection/mixins/minecraft/gui/MixinPlayerListHud.java

```diff
@@ -19,16 +19,120 @@
+    }
+
+    @Inject(method = "render", at = @At(value = "INVOKE", target = "Ljava/lang/Math;min(II)I", shift = At.Shift.BEFORE))
+    private void hookTabColumnHeight(CallbackInfo ci, @Local(ordinal = 5) LocalIntRef o, @Local(ordinal = 6)LocalIntRef p) {
+        if (!ModuleBetterTab.INSTANCE.getRunning()) {
+            return;
+        }
+
+        int totalPlayers = collectPlayerEntries().size();
+        int columns = 1;
+        int rows = totalPlayers;
+
+        while (rows > ModuleBetterTab.Limits.INSTANCE.getHeight()) {
+            columns++;
+            rows = (totalPlayers + columns - 1) / columns;
+        }
+
+        o.set(rows);
+        p.set(columns);
+    }
+
+    @ModifyArg(method = "render", at = @At(value = "INVOKE", target = "Ljava/lang/Math;min(II)I"), index = 0)
+    private int hookWidth(int width) {
+        return ModuleBetterTab.INSTANCE.getRunning() && ModuleBetterTab.AccurateLatency.INSTANCE.getRunning() ? width + 30 : width;
+    }
+
+    @Inject(method = "renderLatencyIcon", at = @At("HEAD"), cancellable = true)
+    private void hookOnRenderLatencyIcon(DrawContext context, int width, int x, int y, PlayerListEntry entry, CallbackInfo ci) {
+        var accurateLatency = ModuleBetterTab.AccurateLatency.INSTANCE;
+        if (ModuleBetterTab.INSTANCE.getRunning() && accurateLatency.getRunning()) {
+            TextRenderer textRenderer = MinecraftClient.getInstance().textRenderer;
+
+            int latency = MathHelper.clamp(entry.getLatency(), 0, 9999);
+            int color = latency < 150 ? 0x00E970 : latency < 300 ? 0xE7D020 : 0xD74238;
+            String text = latency + (accurateLatency.getSuffix() ? "ms" : "");
+            context.drawTextWithShadow(textRenderer, text, x + width - textRenderer.getWidth(text), y, color);
+            ci.cancel();
+        }
+    }
+
+    @ModifyArgs(method = "render", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/DrawContext;fill(IIIII)V", ordinal = 2))
+    private void hookRenderPlayerBackground(Args args, @Local(ordinal = 13) int w, @Local(ordinal = 0) List<PlayerListEntry> entries) {
+        if (!ModuleBetterTab.INSTANCE.getRunning()) {
+            return;
+        }
+
+        var highlight = ModuleBetterTab.Highlight.INSTANCE;
+        if (!highlight.getRunning()) {
+            return;
+        }
+
+        if (w < entries.size()) {
+            var entry = entries.get(w);
+            if (highlight.getSelf().getRunning()) {
+                if (Objects.equals(entry.getProfile().getName(), MinecraftClient.getInstance().player.getGameProfile().getName())) {
+                    args.set(4, highlight.getSelf().getColor().toARGB());
+                    return;
+                }
+            }
+
+            if (highlight.getFriends().getRunning()) {
+                for(var friend : FriendManager.INSTANCE.getFriends()) {
+                    if (Objects.equals(entry.getProfile().getName(), friend.getName())) {
```

Comment:
Is it really necessary to check if an entry is a friend / yourself every frame? Can't we cache this data?

Reply:
how?

Reply:
You could create a field in the `PlayerListEntry`.

Reply:
help me? i I don't understand what you mean

Reply:
resolved